source/tutorials/source-file-selection.md

```diff
@@ -0,0 +1,645 @@
+runCommand "hello" { } ''
+  echo hello world
+''
+```
+
+And try it out:
+```shell-session
+$ nix-build -A package
+this derivation will be built:
+  /nix/store/kmf9sw8fn7ps3ndqs31hvqwsa35b8l3g-hello.drv
+building '/nix/store/kmf9sw8fn7ps3ndqs31hvqwsa35b8l3g-hello.drv'...
+hello world
+error: builder for '/nix/store/kmf9sw8fn7ps3ndqs31hvqwsa35b8l3g-hello.drv'
+  failed to produce output path for output 'out'
+```
+
+We could also add `touch $out` to make the build succeed,
+but we'll omit that for the sake of the tutorial, since we only need the build logs.
+This also makes it easier to build it again, since successful derivation builds would get cached.
+From now on we'll also make build outputs a bit shorter for the sake of brevity.
+
+## Builtins
+
+### Coercion of paths to strings
+
+The easiest way to use local files in builds is using the built-in coercion of [paths](https://nixos.org/manual/nix/stable/language/values.html#type-path) to strings.
+
+Let's create a local `string.txt` file:
+```
+$ echo "This is a string" > string.txt
+```
+
+The two main ways to coerce paths to strings are:
+- Interpolating paths in strings. To try that, change your `package.nix` file to:
+  ```nix
+  { runCommand }:
+  runCommand "file-coercion" { } ''
+    (
+      set -x
+      cat ${./string.txt}
+    )
+  ''
+  ```
+
+  :::{note}
+  Interpolation into bash scripts generally requires [`lib.escapeShellArg`](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.strings.escapeShellArg) for correct escaping.
+  In this case however, the interpolation results in a Nix store path of the form `/nix/store/<hash>-<name>`,
+  and all valid characters of such store paths don't need to be escaped in bash.
+  :::
+
+- Using paths as derivation attributes. To try that, change your `package.nix` file to:
+  ```nix
+  { runCommand }:
+  runCommand "file-coercion" {
+    stringFile = ./string.txt;
+  } ''
+    (
+      set -x
+      cat "$stringFile"
+    )
+  ''
+  ```
+
+  :::{note}
+  Nowadays using the explicit `env` attribute is recommended to set environment variables.
+  `env` doesn't implicitly coerce paths to strings, so it requires using string intepolation instead:
+  ```nix
+  { runCommand }:
+  runCommand "file-coercion" {
+    env.stringFile = "${./string.txt}";
+  } ''
+    (
+      set -x
+      cat "$stringFile"
+    )
+  ''
+  ```
+  :::
+
+These all do the same when built:
+```shell-session
+$ nix-build -A package
+building '/nix/store/9fi0khrkmqw5srjzjsfa0b05hf8div4c-file-coercion.drv'...
+++ cat /nix/store/j5lwpnlfrngks3bpidfr5hcrhgq0fy78-string.txt
+This is a string
+```
+
+As you can see, the `string.txt` file was hashed and added to the store,
+which then allowed the build to access it.
+
+The underlying functionality is the same as `nix-store --add` on an absolute path:
+```shell-session
+$ nix-store --add $(pwd)/string.txt
+/nix/store/j5lwpnlfrngks3bpidfr5hcrhgq0fy78-string.txt
+```
+
+### Coercion of directory paths to strings
```

Comment:
@fricklerhandwerk This one is specific to directories, indeed this one should be replaced with `lib.fileset`. I'm okay removing this section, though ideally it should be _somewhere_, because it's very common to see this in the wild, people should understand how this mechanism works, even if they shouldn't use it themselves. And unless we deprecate this in Nix itself, people won't stop using it.